File: PatTool_Back-End/src/main/java/com/pat/service/MigrationService.java
package com.pat.service;

import com.pat.repo.EvenementsRepository;
import com.pat.repo.domain.Evenement;
import com.pat.repo.domain.UrlEvent;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Date;
import java.util.List;

/**
 * Service for handling data migrations
 * Created for PatTool application
 */
@Service
public class MigrationService {

    private static final Logger log = LoggerFactory.getLogger(MigrationService.class);

    @Autowired
    private EvenementsRepository evenementsRepository;

    /**
     * Migrate photosUrl field to urlEvents for all events
     * @return Migration result message
     */
    public String migratePhotosToUrlEvents() {
        log.info("Starting migration of photosUrl to urlEvents");
        
        List<Evenement> events = evenementsRepository.findAll();
        int migratedCount = 0;
        
        for (Evenement event : events) {
            if (event.getPhotosUrl() != null && !event.getPhotosUrl().isEmpty()) {
                // Initialize urlEvents list if null
                if (event.getUrlEvents() == null) {
                    event.setUrlEvents(new java.util.ArrayList<>());
                }
                
                // Create UrlEvent for each photo URL
                for (String photoUrl : event.getPhotosUrl()) {
                    if (photoUrl != null && !photoUrl.trim().isEmpty()) {
                        UrlEvent photoUrlEvent = new UrlEvent(
                            "Photos",                       // typeUrl
                            new Date(),                     // dateCreation
                            "Patricou",                     // owner
                            photoUrl.trim(),                // link
                            "Photos"                        // urlDescription
                        );
                        event.getUrlEvents().add(photoUrlEvent);
                    }
                }
                
                // Clear the photosUrl field
                event.setPhotosUrl(new java.util.ArrayList<>());
                
                // Save the updated event
                evenementsRepository.save(event);
                migratedCount++;
                
                log.info("Migrated photos for event: {} - {} photos migrated", 
                    event.getEvenementName(), event.getUrlEvents().size());
            }
        }
        
        String result = String.format("Migration des photos terminée avec succès. %d événements migrés.", migratedCount);
        log.info(result);
        return result;
    }

    /**
     * Migrate map field to urlEvents for all events
     * @return Migration result message
     */
    public String migrateMapToUrlEvents() {
        log.info("Starting migration of map to urlEvents");
        
        List<Evenement> events = evenementsRepository.findAll();
        int migratedCount = 0;
        
        for (Evenement event : events) {
            if (event.getMap() != null && !event.getMap().trim().isEmpty()) {
                // Initialize urlEvents list if null
                if (event.getUrlEvents() == null) {
                    event.setUrlEvents(new java.util.ArrayList<>());
                }
                
                // Create UrlEvent from map data
                UrlEvent mapUrlEvent = new UrlEvent(
                    "MAP",                           // typeUrl
                    new Date(),                      // dateCreation
                    "Patricou",                      // owner
                    event.getMap().trim(),           // link
                    "Carte"                         // urlDescription
                );
                
                // Add the UrlEvent to the list
                event.getUrlEvents().add(mapUrlEvent);
                
                // Clear the map field
                event.setMap("");
                
                // Save the updated event
                evenementsRepository.save(event);
                migratedCount++;
                
                log.info("Migrated map for event: {} - map URL: {}", 
                    event.getEvenementName(), mapUrlEvent.getLink());
            }
        }
        
        String result = String.format("Migration des cartes terminée avec succès. %d événements migrés.", migratedCount);
        log.info(result);
        return result;
    }

    /**
     * Get migration status for photos
     * @return Status message
     */
    public String getPhotosMigrationStatus() {
        List<Evenement> events = evenementsRepository.findAll();
        int totalEvents = events.size();
        int eventsWithPhotos = 0;
        int eventsWithUrlEvents = 0;
        
        for (Evenement event : events) {
            if (event.getPhotosUrl() != null && !event.getPhotosUrl().isEmpty()) {
                eventsWithPhotos++;
            }
            if (event.getUrlEvents() != null && !event.getUrlEvents().isEmpty()) {
                eventsWithUrlEvents++;
            }
        }
        
        return String.format(
            "Statut de la migration des photos:\n" +
            "Total événements: %d\n" +
            "Événements avec photosUrl: %d\n" +
            "Événements avec urlEvents: %d\n" +
            "Événements prêts pour migration des photos: %d",
            totalEvents, eventsWithPhotos, eventsWithUrlEvents, eventsWithPhotos
        );
    }

    /**
     * Get migration status for maps
     * @return Status message
     */
    public String getMapMigrationStatus() {
        List<Evenement> events = evenementsRepository.findAll();
        int totalEvents = events.size();
        int eventsWithMap = 0;
        int eventsWithUrlEvents = 0;
        
        for (Evenement event : events) {
            if (event.getMap() != null && !event.getMap().trim().isEmpty()) {
                eventsWithMap++;
            }
            if (event.getUrlEvents() != null && !event.getUrlEvents().isEmpty()) {
                eventsWithUrlEvents++;
            }
        }
        
        return String.format(
            "Statut de la migration des cartes:\n" +
            "Total événements: %d\n" +
            "Événements avec champ map: %d\n" +
            "Événements avec urlEvents: %d\n" +
            "Événements prêts pour migration: %d",
            totalEvents, eventsWithMap, eventsWithUrlEvents, eventsWithMap
        );
    }
}

File: PatTool_Front-End/src/app/services/migration.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable, from, forkJoin } from 'rxjs';
import { map, switchMap } from 'rxjs/operators';
import { environment } from '../../environments/environment';
import { KeycloakService } from '../keycloak/keycloak.service';
import { Evenement } from '../model/evenement';
import { UrlEvent } from '../model/url-event';

@Injectable({
  providedIn: 'root'
})
export class MigrationService {
  private apiUrl = environment.API_URL;

  constructor(private http: HttpClient, private keycloakService: KeycloakService) { }

  // Get the header with token for Keycloak Security
  private getHeaderWithToken(): Observable<HttpHeaders> {
    return from(this.keycloakService.getToken()).pipe(
      map(token => new HttpHeaders({
        'Accept': 'application/json',
        'Content-Type': 'application/json',
        'Authorization': 'Bearer ' + token
      }))
    );
  }

  /**
   * Execute the migration from map field to urlEvents using existing APIs
   */
  migrateMapToUrlEventsFrontend(): Observable<string> {
    return this.getHeaderWithToken().pipe(
      switchMap(headers => {
        // Get all events
        return this.http.get<Evenement[]>(`${this.apiUrl}evenements`, { headers }).pipe(
          switchMap(events => {
            const eventsToMigrate = events.filter(event => 
              event.map && event.map.trim() !== ''
            );
            
            if (eventsToMigrate.length === 0) {
              return from(['Aucun événement à migrer trouvé.']);
            }

            // Create migration operations for each event
            const migrationOperations = eventsToMigrate.map(event => {
              // Create UrlEvent from map data
              const mapUrlEvent = new UrlEvent(
                "MAP",                           // typeUrl: Use the new ID constant
                new Date(),                      // dateCreation: Current date
                "Patricou",                      // owner: As requested
                event.map.trim(),                // link: The map URL
                "Carte"                         // urlDescription: As requested
              );
              
              // Initialize urlEvents list if null
              if (!event.urlEvents) {
                event.urlEvents = [];
              }
              
              // Add the UrlEvent to the list
              event.urlEvents.push(mapUrlEvent);
              
              // Clear the map field
              event.map = "";
              
              // Update the event
              return this.http.put<Evenement>(`${this.apiUrl}evenements/${event.id}`, event, { headers });
            });

            // Execute all migrations
            return forkJoin(migrationOperations).pipe(
              map(() => `Migration terminée avec succès. ${eventsToMigrate.length} événements migrés.`)
            );
          })
        );
      })
    );
  }

  /**
   * Get migration status using existing APIs
   */
  getMigrationStatusFrontend(): Observable<string> {
    return this.getHeaderWithToken().pipe(
      switchMap(headers =>
        this.http.get<Evenement[]>(`${this.apiUrl}evenements`, { headers }).pipe(
          map(events => {
            const eventsWithMap = events.filter(event => 
              event.map && event.map.trim() !== ''
            ).length;
            const eventsWithUrlEvents = events.filter(event => 
              event.urlEvents && event.urlEvents.length > 0
            ).length;
            const totalEvents = events.length;

            return `Statut de la migration:\n` +
                   `Total événements: ${totalEvents}\n` +
                   `Événements avec champ map: ${eventsWithMap}\n` +
                   `Événements avec urlEvents: ${eventsWithUrlEvents}\n` +
                   `Événements prêts pour migration: ${eventsWithMap}`;
          })
        )
      )
    );
  }

  /**
   * Execute the migration from map field to urlEvents (backend endpoint)
   */
  migrateMapToUrlEvents(): Observable<string> {
    return this.getHeaderWithToken().pipe(
      switchMap(headers =>
        this.http.post<string>(`${this.apiUrl}migration/migrate-map-to-urlevents`, {}, { headers })
      )
    );
  }

  /**
   * Get the current migration status (backend endpoint)
   */
  getMigrationStatus(): Observable<string> {
    return this.getHeaderWithToken().pipe(
      switchMap(headers =>
        this.http.get<string>(`${this.apiUrl}migration/migration-status`, { headers })
      )
    );
  }

  /**
   * Execute the migration from photosUrl field to urlEvents using existing APIs
   */
  migratePhotosToUrlEventsFrontend(): Observable<string> {
    return this.getHeaderWithToken().pipe(
      switchMap(headers => {
        // Get all events
        return this.http.get<Evenement[]>(`${this.apiUrl}evenements`, { headers }).pipe(
          switchMap(events => {
            const eventsToMigrate = events.filter(event => 
              event.photosUrl && event.photosUrl.length > 0
            );
            
            if (eventsToMigrate.length === 0) {
              return from(['Aucun événement avec photos à migrer trouvé.']);
            }

            // Create migration operations for each event
            const migrationOperations = eventsToMigrate.map(event => {
              // Initialize urlEvents list if null
              if (!event.urlEvents) {
                event.urlEvents = [];
              }
              
              // Create UrlEvent for each photo URL
              const photoUrlEvents = event.photosUrl.map(photoUrl => 
                new UrlEvent(
                  "Photos",                       // typeUrl: Corrected to "Photos"
                  new Date(),                     // dateCreation: Current date
                  "Patricou",                     // owner: As requested
                  photoUrl.trim(),                // link: The photo URL
                  "Photos"                        // urlDescription: Corrected to "Photos"
                )
              );
              
              // Add all photo UrlEvents to the list
              event.urlEvents.push(...photoUrlEvents);
              
              // Clear the photosUrl field
              event.photosUrl = [];
              
              // Update the event
              return this.http.put<Evenement>(`${this.apiUrl}evenements/${event.id}`, event, { headers });
            });

            // Execute all migrations
            return forkJoin(migrationOperations).pipe(
              map(() => `Migration des photos terminée avec succès. ${eventsToMigrate.length} événements migrés.`)
            );
          })
        );
      })
    );
  }

  /**
   * Get migration status for photos using existing APIs
   */
  getPhotosMigrationStatusFrontend(): Observable<string> {
    return this.getHeaderWithToken().pipe(
      switchMap(headers =>
        this.http.get<Evenement[]>(`${this.apiUrl}evenements`, { headers }).pipe(
          map(events => {
            const eventsWithPhotos = events.filter(event => 
              event.photosUrl && event.photosUrl.length > 0
            ).length;
            const eventsWithUrlEvents = events.filter(event => 
              event.urlEvents && event.urlEvents.length > 0
            ).length;
            const totalEvents = events.length;

            return `Statut de la migration des photos:\n` +
                   `Total événements: ${totalEvents}\n` +
                   `Événements avec photosUrl: ${eventsWithPhotos}\n` +
                   `Événements avec urlEvents: ${eventsWithUrlEvents}\n` +
                   `Événements prêts pour migration des photos: ${eventsWithPhotos}`;
          })
        )
      )
    );
  }

  /**
   * Execute the migration from photosUrl field to urlEvents (backend endpoint)
   */
  migratePhotosToUrlEvents(): Observable<string> {
    return this.getHeaderWithToken().pipe(
      switchMap(headers =>
        this.http.post<string>(`${this.apiUrl}migration/migrate-photos-to-urlevents`, {}, { headers })
      )
    );
  }

  /**
   * Get the current migration status for photos (backend endpoint)
   */
  getPhotosMigrationStatus(): Observable<string> {
    return this.getHeaderWithToken().pipe(
      switchMap(headers =>
        this.http.get<string>(`${this.apiUrl}migration/photos-migration-status`, { headers })
      )
    );
  }
}
